TaskManagementSystem/Services/TasksService.cs

```diff
@@ -15,117 +15,134 @@ namespace TaskManagementSystem.Services;
+	
+        if (updatedRows is 0)
+            throw new NotFoundException(ErrorMessageConstants.LockedTaskWithIdDoesNotExist);
     }
 
-    private async Task<TaskEntity> GetTaskEntityAsync(int taskId)
+    public async Task<DeleteAction> DeleteTaskAsync(int taskId)
     {
-        var taskEntity = await _dbContext.Tasks.FindAsync(taskId) 
+        var taskEntity = await _dbContext.Tasks.FindAsync(taskId)
             ?? throw new NotFoundException(ErrorMessageConstants.TaskDoesNotExist);
 
-        return taskEntity;
+        ValidateTaskStatusForDeletion(taskEntity);
+
+        var deleteAction = GetDeleteAction(taskEntity.Priority);
+
+        await _taskDeleteContext.HandleAsync(taskEntity, deleteAction);
+
+        return deleteAction;
     }
 
-    private async Task ValidatateCategoryExists(int categoryId)
+    private async Task ValidatateCategoryExistsAsync(int categoryId)
     {
-        var categoryExists = await _categoriesService.CategoryExistsAsync(categoryId);
+        var categoryExists = await _categoryRepository.CategoryExistsAsync(categoryId);
 
         if (!categoryExists)
             throw new BadHttpRequestException(ErrorMessageConstants.CategoryDoesNotExist);
     }
 
-    private bool IsCompleted(Status taskStatus)
+    private static void ValidateTaskStatusForDeletion(TaskEntity taskEntity)
     {
-        return taskStatus == Status.Completed || taskStatus == Status.Archived
-            ? true
-            : false;
+        if (taskEntity.Status == Status.Locked)
+            throw new ConflictException(ErrorMessageConstants.TaskAlreadyLocked);
+
+        if (taskEntity.Status == Status.Archived)
+            throw new ConflictException(ErrorMessageConstants.ArchivedTaskCanNotBeDeleted);
     }
 
-    private async Task ValidateTaskDataForUpdate(TaskEntity taskEntity, UpdateTaskRequestDto taskDto)
+    private void ValidateTaskStatusForUpdateAsync(TaskEntity taskEntity, UpdateTaskRequestDto taskDto)
     {
         var canArchivedBeEdited = taskEntity.Status == Status.Archived;
         if (canArchivedBeEdited)
-            throw new BadHttpRequestException(ErrorMessageConstants.ArchivedTaskCanNotBeEdited);
+            throw new ConflictException(ErrorMessageConstants.ArchivedTaskCanNotBeEdited);
 
         var canArchiveEntity = taskEntity.Status != Status.Completed && taskDto.Status == Status.Archived;
         if (canArchiveEntity)
-            throw new BadHttpRequestException(ErrorMessageConstants.OnlyCompletedTaskCanBeArchived);
+            throw new ConflictException(ErrorMessageConstants.OnlyCompletedTaskCanBeArchived);
 
-        await ValidatateCategoryExists(taskDto.CategoryId);
+        var canLockedBeEdited = taskEntity.Status == Status.Locked;
+        if (canLockedBeEdited)
+            throw new ConflictException(ErrorMessageConstants.LockedTaskCanNotBeEdited);
     }
-}
+
+    private DeleteAction GetDeleteAction(Priority priority) => priority switch
```

Comment:
This logic can be removed from the service. Instead pass only the task to the context.